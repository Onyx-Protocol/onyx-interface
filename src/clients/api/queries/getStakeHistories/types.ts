import { StakeHistory } from 'types';

export interface GetStakeHistoriesInput {
  limit?: number;
  page?: number;
  address?: string;
}

export interface StakeHistoryApiResponse {
  id: number;
  address: string;
  tx_hash: string;
  block_hash: string;
  amount: string;
  price: string;
  reward: string;
  type: number;
  created_at: string;
  updated_at: string;
}

export interface StakeHistoriesApiResponse {
  data: StakeHistoryApiResponse[];
  metadata: {
    page: number;
    limit: number;
    totalItem: number;
    totalPage: number;
  };
}

export interface GetStakeHistoriesOutput {
  page: number;
  limit: number;
  total: number;
  stakeHistories: StakeHistory[];
}
