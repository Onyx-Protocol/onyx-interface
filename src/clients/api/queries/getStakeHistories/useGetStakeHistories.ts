import { QueryObserverOptions, useQuery } from 'react-query';

import getStakeHistories, {
  GetStakeHistoriesInput,
  GetStakeHistoriesOutput,
} from 'clients/api/queries/getStakeHistories';
import { BLOCK_TIME_MS } from 'constants/ethereum';
import FunctionKey from 'constants/functionKey';

type Options = QueryObserverOptions<
  GetStakeHistoriesOutput,
  Error,
  GetStakeHistoriesOutput,
  GetStakeHistoriesOutput,
  [FunctionKey.GET_STAKE_HISTORIES, GetStakeHistoriesInput]
>;

const useGetStakeHistories = (params: GetStakeHistoriesInput = {}, options?: Options) =>
  // This endpoint is paginated so we keep the previous responses by default to
  // create a more seamless paginating experience
  useQuery([FunctionKey.GET_STAKE_HISTORIES, params], () => getStakeHistories(params), {
    keepPreviousData: true,
    placeholderData: { limit: 0, total: 0, page: 0, stakeHistories: [] },
    refetchInterval: params.page === 1 ? BLOCK_TIME_MS * 5 : undefined,
    ...options,
  });

export default useGetStakeHistories;
