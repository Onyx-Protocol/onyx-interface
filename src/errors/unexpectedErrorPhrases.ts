import { t } from 'translation';

export const unexpectedErrorPhrases = {
  somethingWentWrong: t('errors.somethingWentWrong'),
  somethingWentWrongRetrievingTransactions: t('errors.somethingWentWrongRetrievingTransactions'),
  somethingWentWrongRetrievingVoterAccounts: t('errors.somethingWentWrongRetrievingVoterAccounts'),
  somethingWentWrongRetrievingSignedMessage: t('errors.somethingWentWrongRetrievingSignedMessage'),
  somethingWentWrongRetrievingUserInfo: t('errors.somethingWentWrongRetrievingUserInfo'),
  somethingWentWrongRetrievingVoterDetails: t('errors.somethingWentWrongRetrievingVoterDetails'),
  somethingWentWrongRetrievingVoterHistory: t('errors.somethingWentWrongRetrievingVoterHistory'),
  walletNotConnected: t('errors.walletNotConnected'),
  undefinedAccountErrorMessage: t('errors.undefinedAccountErrorMessage'),
  internalErrorXcnToVrtConversionRatioUndefined: t(
    'convertVrt.internalErrorXcnToVrtConversionRatioUndefined',
  ),
  accountError: t('markets.errors.accountError'),
  incorrectSwapInput: t('swap.errors.incorrectSwapInput'),
};
