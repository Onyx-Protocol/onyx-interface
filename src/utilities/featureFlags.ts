// import { TOKENS } from 'constants/tokens';

export const DISABLED_TOKENS = [];

export const isAssetEnabled = (assetId: string) => {
  if (assetId) return true;
};
