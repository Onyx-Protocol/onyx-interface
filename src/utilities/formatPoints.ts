export const formatPoint = (point: number, digit?: number) => {
  if (digit) {
    return (+point).toLocaleString('en-US', {
      maximumFractionDigits: digit || 2,
    });
  }

  let [real, imaginary] = point.toString().split('.');
  real = (+real).toLocaleString('en-US');
  imaginary = imaginary ? `.${imaginary}` : '';
  return `${real}${imaginary}`;
};
