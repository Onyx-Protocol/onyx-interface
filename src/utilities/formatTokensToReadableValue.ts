import BigNumber from 'bignumber.js';
import { Token } from 'types';

import PLACEHOLDER_KEY from 'constants/placeholderKey';

import { shortenValueWithSuffix } from './shortenValueWithSuffix';

export const formatTokensToReadableValue = ({
  value,
  token,
  minimizeDecimals = false,
  shortenLargeValue = false,
  addSymbol = true,
  removeDecimals = false,
}: {
  value: BigNumber | undefined;
  token: Token;
  minimizeDecimals?: boolean;
  shortenLargeValue?: boolean;
  addSymbol?: boolean;
  removeDecimals?: boolean;
}) => {
  if (value === undefined) {
    return PLACEHOLDER_KEY;
  }

  let decimalPlaces;
  if (minimizeDecimals) {
    decimalPlaces = 4;
  } else {
    decimalPlaces = token.decimals;
  }

  if (removeDecimals) {
    decimalPlaces = 0;
  }

  let symbolPlacement = '';
  if (addSymbol) {
    symbolPlacement = ` ${token.symbol}`;
  }

  if (shortenLargeValue) {
    return `${shortenValueWithSuffix({
      value,
    })}${symbolPlacement}`;
  }

  return `${value.dp(decimalPlaces).toFormat()}${symbolPlacement}`;
};

export default formatTokensToReadableValue;
