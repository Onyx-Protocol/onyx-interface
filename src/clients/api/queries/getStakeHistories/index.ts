import { VError } from 'errors';
import { formatToStakeHistory, restService } from 'utilities';

import {
  GetStakeHistoriesInput,
  GetStakeHistoriesOutput,
  StakeHistoriesApiResponse,
} from './types';

export * from './types';

const getStakeHistories = async ({
  page = 1,
  limit = 6,
  address,
}: GetStakeHistoriesInput): Promise<GetStakeHistoriesOutput> => {
  if (!address) {
    return { stakeHistories: [], limit, total: 0, page };
  }
  const response = await restService<StakeHistoriesApiResponse>({
    endpoint: '/staking/history',
    method: 'GET',
    params: { page: page + 1, limit, address },
    gov: true,
  });
  const payload = response.data?.data;

  // @todo Add specific api error handling
  if ('result' in response && response.result === 'error') {
    throw new VError({
      type: 'unexpected',
      code: 'somethingWentWrong',
      data: { message: response.message },
    });
  }

  if (!payload) {
    throw new VError({ type: 'unexpected', code: 'somethingWentWrong' });
  }

  const { limit: payloadLimit, totalItem: total, page: payloadPage } = payload.metadata;
  const stakeHistories = payload.data.map(p => formatToStakeHistory(p));

  return { stakeHistories, limit: payloadLimit, total, page: payloadPage };
};

export default getStakeHistories;
