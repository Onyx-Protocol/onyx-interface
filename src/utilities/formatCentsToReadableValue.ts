/* eslint-disable no-nested-ternary */
import BigNumber from 'bignumber.js';
import { shortenValueWithSuffix } from 'utilities';

import PLACEHOLDER_KEY from 'constants/placeholderKey';

const formatCentsToReadableValue = ({
  value,
  shortenLargeValue = false,
}: {
  value: number | BigNumber | undefined;
  shortenLargeValue?: boolean;
}) => {
  if (value === undefined) {
    return PLACEHOLDER_KEY;
  }

  const wrappedValueDollars = new BigNumber(value).dividedBy(100);

  if (!shortenLargeValue) {
    return `$${
      wrappedValueDollars.gt(0.01)
        ? wrappedValueDollars.toFormat(2)
        : wrappedValueDollars.gt(0.001)
        ? wrappedValueDollars.toFormat(3)
        : wrappedValueDollars.gt(0.0001)
        ? wrappedValueDollars.toFormat(4)
        : wrappedValueDollars.toFormat(5)
    }`;
  }

  // Shorten value
  const shortenedValue = shortenValueWithSuffix({
    value: wrappedValueDollars,
    outputsDollars: true,
  });

  return `$${shortenedValue}`;
};

export default formatCentsToReadableValue;
