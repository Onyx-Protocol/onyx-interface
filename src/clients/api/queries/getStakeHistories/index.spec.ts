import { VError } from 'errors';
import { restService } from 'utilities';

import proposalResponse from '__mocks__/api/stakeHistories.json';

import getStakeHistories from '.';

jest.mock('utilities/restService');

describe('api/queries/getProposals', () => {
  test('throws an error when request fails', async () => {
    const fakeErrorMessage = 'Fake error message';

    (restService as jest.Mock).mockImplementationOnce(async () => ({
      result: 'error',
      status: false,
      message: fakeErrorMessage,
    }));

    try {
      await getStakeHistories({ address: '0x0000000000000000000000000000000000000000' });

      throw new Error('getProposals should have thrown an error but did not');
    } catch (error) {
      expect(error).toBeInstanceOf(VError);
      if (error instanceof VError) {
        expect(error.type).toBe('unexpected');
        expect(error.code).toBe('somethingWentWrong');
        expect(error.message).toBe('somethingWentWrong');
        expect(error.data.message).toBe(fakeErrorMessage);
      }
    }
  });

  test('returns formatted proposals', async () => {
    (restService as jest.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: { data: proposalResponse, limit: 20, offset: 20 },
    }));

    const response = await getStakeHistories({
      limit: 10,
      page: 2,
      address: '0x0000000000000000000000000000000000000000',
    });

    expect(restService).toBeCalledWith({
      endpoint: '/proposals',
      method: 'GET',
      params: {
        limit: 10,
        offset: 20,
        version: 'v2',
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('Gets called with correct default arguments', async () => {
    (restService as jest.Mock).mockImplementationOnce(async () => ({
      status: 200,
      data: { data: proposalResponse },
    }));

    const response = await getStakeHistories({
      address: '0x0000000000000000000000000000000000000000',
    });

    expect(restService).toBeCalledWith({
      endpoint: '/proposals',
      method: 'GET',
      params: {
        limit: 5,
        offset: 0,
        version: 'v2',
      },
    });

    expect(response).toMatchSnapshot();
  });
});
