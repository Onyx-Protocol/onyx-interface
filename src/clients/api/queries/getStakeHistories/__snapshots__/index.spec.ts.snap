// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/queries/getProposals Gets called with correct default arguments 1`] = `
Object {
  "limit": 5,
  "page": 1,
  "stakeHistories": Array [
    Object {
      "address": "0x0000000000000000000000000000000000000000",
      "amount": "11075000000000000000",
      "block_hash": "0x7d6991ef8f031e9cdc4ced2f924dee008e505c741f169101124aee16dad795f1",
      "created_at": 2022-12-14T19:04:11.000Z,
      "id": 291,
      "price": "0.033695332045177356",
      "reward": "0",
      "tx_hash": "0xf97a7673403a643ad23edb9ffa4c60bf4192998aaf436782240b84be94d3aba1",
      "type": 0,
      "updated_at": 2022-12-14T19:04:11.000Z,
    },
  ],
  "total": 1,
}
`;

exports[`api/queries/getProposals returns formatted proposals 1`] = `
Object {
  "limit": 5,
  "page": 1,
  "stakeHistories": Array [
    Object {
      "address": "0x0000000000000000000000000000000000000000",
      "amount": "11075000000000000000",
      "block_hash": "0x7d6991ef8f031e9cdc4ced2f924dee008e505c741f169101124aee16dad795f1",
      "created_at": 2022-12-14T19:04:11.000Z,
      "id": 291,
      "price": "0.033695332045177356",
      "reward": "0",
      "tx_hash": "0xf97a7673403a643ad23edb9ffa4c60bf4192998aaf436782240b84be94d3aba1",
      "type": 0,
      "updated_at": 2022-12-14T19:04:11.000Z,
    },
  ],
  "total": 1,
}
`;

exports[`api/queries/getStakeHistories Gets called with correct default arguments 1`] = `
Object {
  "limit": 6,
  "page": 1,
  "stakeHistories": Array [
    Object {
      "address": "0xca343afaf6f9193e232e46a074b481aeea548e59",      
      "amount": "11071000000000000000",
      "block_hash": "0x7d6991ef8f031e9cdc4ced2f924dee008e505c741f169101124aee16dad795f0",
      "created_at": "2022-12-14T19:04:11.000Z",
      "id": 291,
      "price": "0.033695332045177356",
      "reward": "0",
      "tx_hash": "0xf97a7673403a643ad23edb9ffa4c60bf4192998aaf436782240b84be94d3aba2",
      "type": 0,
      "updated_at": "2022-12-14T19:04:11.000Z"
    }
  ],
  "total": 1,
}
`;

exports[`api/queries/getStakeHistories returns formatted history 1`] = `
Object {
  "limit": 6,
  "page": 1,
  "stakeHistories": Array [
    Object {
      "address": "0xca343afaf6f9193e232e46a074b481aeea548e59",      
      "amount": "11071000000000000000",
      "block_hash": "0x7d6991ef8f031e9cdc4ced2f924dee008e505c741f169101124aee16dad795f0",
      "created_at": "2022-12-14T19:04:11.000Z",
      "id": 291,
      "price": "0.033695332045177356",
      "reward": "0",
      "tx_hash": "0xf97a7673403a643ad23edb9ffa4c60bf4192998aaf436782240b84be94d3aba2",
      "type": 0,
      "updated_at": "2022-12-14T19:04:11.000Z"
    }
  ],
  "total": 1,
}
`;
